contracts/lib/LibPrices.sol

```diff
@@ -96,48 +134,45 @@ library Prices {
         PriceInstant[24] memory tracerPrices,
         PriceInstant[24] memory oraclePrices
     ) public pure returns (TWAP memory) {
-        uint256 instantDerivative = 0;
+        uint256 totalDerivativeTimeWeight = 0;
+        uint256 totalUnderlyingTimeWeight = 0;
         uint256 cumulativeDerivative = 0;
-        uint256 instantUnderlying = 0;
         uint256 cumulativeUnderlying = 0;
 
         for (uint256 i = 0; i < 8; i++) {
             uint256 currTimeWeight = 8 - i;
             // if hour < i loop back towards 0 from 23.
             // otherwise move from hour towards 0
-            uint256 j = hour < i ? 23 - i + hour : hour - i;
+            uint256 j = hour < i ? 24 - i + hour : hour - i;
 
             uint256 currDerivativePrice = averagePrice(tracerPrices[j]);
             uint256 currUnderlyingPrice = averagePrice(oraclePrices[j]);
 
-            if (currDerivativePrice > 0) {
-                instantDerivative += currTimeWeight;
+            // don't include periods that have no trades
+            if (tracerPrices[j].trades == 0) {
+                continue;
+            } else {
+                totalDerivativeTimeWeight += currTimeWeight;
                 cumulativeDerivative += currTimeWeight * currDerivativePrice;
             }
 
-            if (currUnderlyingPrice > 0) {
-                instantUnderlying += currTimeWeight;
-                cumulativeUnderlying += currTimeWeight * currUnderlyingPrice;
-            }
-
-            if (instantDerivative == 0) {
-                return TWAP(0, 0);
+            // don't include periods that have no trades
+            if (oraclePrices[j].trades == 0) {
+                continue;
             } else {
-                return
-                    TWAP(
-                        PRBMathUD60x18.div(
-                            cumulativeUnderlying,
-                            instantUnderlying
-                        ),
-                        PRBMathUD60x18.div(
-                            cumulativeDerivative,
-                            instantDerivative
-                        )
-                    );
+                totalUnderlyingTimeWeight += currTimeWeight;
+                cumulativeUnderlying += currTimeWeight * currUnderlyingPrice;
             }
         }
+
+        return
+            TWAP(
+                cumulativeUnderlying / totalUnderlyingTimeWeight,
```

Comment:
I think this has come up before, but is there no need for WAD divisions here?

Reply:
I would say it should be WAD

Reply:
Discussed in person; doesn't need to be WAD